### ISSUE TITLE
ViewPager Adapters cleanup

### ISSUE THREAD

**Issue by nmilcoff** [MvvmCross/MvvmCross#2440]
### Description :scroll:
We currently expose four ViewPager Fragment Adapters:
- [MvxFragmentStatePagerAdapter](https://github.com/MvvmCross/MvvmCross/blob/develop/MvvmCross-AndroidSupport/MvvmCross.Droid.Support.Fragment/MvxFragmentStatePagerAdapter.cs)
- [MvxFragmentPagerAdapter](https://github.com/MvvmCross/MvvmCross/blob/develop/MvvmCross-AndroidSupport/MvvmCross.Droid.Support.Fragment/MvxFragmentPagerAdapter.cs)
- [MvxCachingFragmentStatePagerAdapter](https://github.com/MvvmCross/MvvmCross/blob/develop/MvvmCross-AndroidSupport/MvvmCross.Droid.Support.Fragment/MvxCachingFragmentStatePagerAdapter.cs)
- [MvxCachingFragmentPagerAdapter](https://github.com/MvvmCross/MvvmCross/blob/develop/MvvmCross-AndroidSupport/MvvmCross.Droid.Support.Fragment/MvxCachingFragmentPagerAdapter.cs)

It looks like the ones with the word "Caching" are there because in 2014 there was an issue with the adapters provided by the Android SDK.

Maybe it's time to remove them?

### Expected behavior :thinking:
We should expose only two ViewPager adapters.

### Actual behavior :bug:
We are exposing too many ViewPager adapters

### Configuration 🔧 

**Version:** 5.x

**Platform:** 
- [ ] :iphone: iOS
- [X] :robot: Android
- [ ] :checkered_flag: WPF
- [ ] :earth_americas: UWP
- [ ] :apple: MacOS
- [ ] :tv: tvOS
- [ ] :monkey: Xamarin.Forms

**Comment by Qwin** [comment]
Seems like 
MvxFragmentStatePagerAdapter
MvxFragmentPagerAdapter
Both dont add support to add Tags on each fragment, while the caching ones do.

**Comment by nmilcoff** [comment]
@Qwin can you elaborate a bit more please? I've been browsing the source code but I don't fully get it. It looks like those adapters only provide getter methods for tags.

**Comment by Qwin** [comment]
@nmilcoff yes of course, if you look at MvxCachingFragmentPagerAdapter.cs you will see this : 
<img width="674" alt="screen shot 2017-12-18 at 6 33 43 pm" src="https://user-images.githubusercontent.com/1026038/34136207-117f63da-e422-11e7-83a4-fdfa041ecc0e.png">
In the the last line you will see they are setting for the fragment a fragmentTag (which is just the name of the fragment type) by using the fragmentmanager. the tag is used for adding nested fragments. Nested fragments use the method GetFragmentByFragmentTag internally. I hope its clear if you want to know about how the nested fragment works internally I can elaborate even more on it.   

Anyways the 
MvxFragmentStatePagerAdapter
MvxFragmentPagerAdapter
both dont add a tag. I am guessing this is just missing ?

**Comment by Qwin** [comment]
@nmilcoff I noticed even though we are adding the Tag in the custom implementation (caching classes), it would still require some work on the: [https://github.com/MvvmCross/MvvmCross/blob/c4b41183eed4f3b800f736afb7482e462a6d7f71/MvvmCross-AndroidSupport/MvvmCross.Droid.Support.V7.AppCompat/MvxAppCompatViewPresenter.cs](MvxAppCompatViewPresenter.cs)
As MvxAppCompatViewPresenter only supports SupportFragmentManager or (if using the non appcompat) FragmentManager but both do not currently support ChildFragmentManager which is the required one to support nested fragments in ViewPagers. 

I would suggest the following happens if we want to do a good job at removing the Caching classes: 
1. Add the tagging by overriding the InstantiateItem in MvxFragmentPagerAdapter 
2. Maybe add support for ChildFragmentManager in the android presenters (and also support for multi level nesting)
3. Remove the Caching ones.

**Comment by Qwin** [comment]
So I think @nmilcoff based on what I have seen so far is that we need do need to override the InstantiateItem for both MvxFragmentStatePagerAdapter and MvxFragmentPagerAdapter  and call ourselves the GetItem as well and set a default tag on it like so:

<img width="802" alt="screen shot 2017-12-19 at 4 35 00 pm" src="https://user-images.githubusercontent.com/1026038/34183952-9a45cf02-e4da-11e7-8807-74f903fc12c1.png">

This will cause each added fragment to have at least a default Tag and when people want to override that Tag they can by overriding the GetTag method. 

Also looking at the presenter I already implemented a way to support ChildFragmentManager (recursive) like so (this will search the hierarchie for the right Tag on a Fragment and can then handle setting it on the view, so we would need to add this to both presenters in Android) : 
<img width="742" alt="screen shot 2017-12-19 at 4 37 18 pm" src="https://user-images.githubusercontent.com/1026038/34184045-09446544-e4db-11e7-8150-f4db54725be4.png">

I tested the code and it works as expected.

**Comment by martijn00** [comment]
@Qwin you are on the right track i think! Can you make a PR so we can test this?

**Comment by Qwin** [comment]
@martijn00 will do :)

**Comment by nmilcoff** [comment]
@Qwin based on your work on #2482 what are we missing to remove the caching adapters?

From your previous comment looks like this is still necessary:

> override the InstantiateItem for both MvxFragmentStatePagerAdapter and MvxFragmentPagerAdapter and call ourselves the GetItem as well and set a default tag

**Comment by nickrandolph** [comment]
@nmilcoff @martijn00 is this still going to make it into 6.0, or can we push to 6.1?

**Comment by nmilcoff** [comment]
@nickrandolph we should take the chance to work on this for v6, as it involves a breaking change.

**Comment by nickrandolph** [comment]
@nmilcoff are we able to generate a PR that demonstrates this issue in the playground, so we can get this issue resolved

**Comment by nmilcoff** [comment]
I've been digging more about this and it seems to me the most reasonable way to cleanup our adapters might be to keep the `Caching` ones and remove the others, because otherwise we  loose support for nested fragments. 

To avoid this we would need to fully override the behavior of `InstantiateItem` method in `MvxFragmentPagerAdapter` and `MvxFragmentStatePagerAdapter`... which is exactly what the `Caching` adapters are doing :)

We must then modify the Android ViewPresenters to support the nesting scenario better, but that's another step.

What do you people think about this? (cc: @Qwin @wh1t3cAt1k)

**Comment by nickrandolph** [comment]
@nmilcoff is this completed now, or is there more work to be done?

**Comment by nmilcoff** [comment]
Well... I would like to have only two ViewPager adapters and not four, but that doesn't bring any actual value to our users - and it is a breaking change. 

So I would say we can schedule it for the next major version.
